src/org/labkey/test/components/ui/grids/GridFilterModal.java

```diff
@@ -0,0 +1,139 @@
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+/**
+ * Wraps 'labkey-ui-component' defined in <code>public/QueryModel/GridFilterModal.tsx</code>
+ */
+public class GridFilterModal extends ModalDialog
+{
+    protected static final Locator listItem = Locator.byClass("list-group-item");
+
+    private final UpdatingComponent _linkedComponent;
+
+    public GridFilterModal(WebDriver driver, UpdatingComponent linkedComponent)
+    {
+        super(new ModalDialogFinder(driver));
+        _linkedComponent = linkedComponent;
+    }
+
+    @Override
+    protected void waitForReady(ModalDialog.ElementCache ec)
+    {
+        getWrapper().shortWait().until(ExpectedConditions.visibilityOf(listItem.findElementOrNull(elementCache().fieldsSelectionPanel)));
+    }
+
+    /**
+     * Select field to configure filters for
+     * @param fieldLabel Field's label
+     * @return this component
+     */
+    public GridFilterModal selectField(String fieldLabel)
+    {
+        WebElement fieldItem = listItem.withText(fieldLabel).findElement(elementCache().fieldsSelectionPanel);
+        fieldItem.click();
+        Locator.byClass("filter-modal__col-sub-title").withText("Find values for " + fieldLabel)
+                .waitForElement(elementCache().filterPanel, 10_000);
+
+        return this;
+    }
+
+    /**
+     * Select the filter expression tab for the current field.
+     * @return panel for configuring the filter expression
+     */
+    public FilterExpressionPanel selectExpressionTab()
+    {
+        elementCache().filterTabs.selectTab("Filter");
+        return elementCache().filterExpressionPanel;
+    }
+
+    /**
+     * Select the facet tab for the current field. Will throw <code>NoSuchElementException</code> if tab isn't present.
+     * @return panel for configuring faceted filter
+     */
+    public FilterFacetedPanel selectFacetTab()
+    {
+        elementCache().filterTabs.selectTab("Choose values");
+        return elementCache().filterFacetedPanel;
+    }
+
+    /**
+     * Save current changes to the search criteria.
+     * Throw <code>IllegalStateException</code> if the save button is disabled because no changes have been made.
+     */
+    public void confirm()
+    {
+        confirm(false);
+    }
+
+    public void confirm(boolean errorExpected)
+    {
+        if (!elementCache().submitButton.isEnabled())
+        {
+            throw new IllegalStateException("Confirmation button is not enabled.");
+        }
+
+        if (errorExpected)
+        {
+            elementCache().submitButton.click();
+            getWrapper().shortWait().until(ExpectedConditions.visibilityOf(elementCache().errorAlert));
+        }
+        else
+        {
+            _linkedComponent.doAndWaitForUpdate(() -> {
+                elementCache().submitButton.click();
+                waitForClose();
+            });
+        }
+    }
```

Comment:
The preferred pattern is to have separate methods for distinct behavior instead of using a flag. In this case it would allow the confirm with error to return the error.
```suggestion
    /**
     * Save current changes to the search criteria.
     * Throw <code>IllegalStateException</code> if the save button is disabled because no changes have been made.
     */
    public void confirm()
    {
        _linkedComponent.doAndWaitForUpdate(() -> {
            clickConfirm();
            waitForClose();
        });
    }

    public String confirmExpectingError()
    {
        clickConfirm();
        return getWrapper().shortWait().until(ExpectedConditions.visibilityOf(elementCache().errorAlert)).getText();
    }

    protected void clickConfirm()
    {
        if (!elementCache().submitButton.isEnabled())
        {
            throw new IllegalStateException("Confirmation button is not enabled.");
        }

        elementCache().submitButton.click();
    }
```

Reply:
Sounds good. Thanks.